tools/header_guard_check/lib/header_guard_check.dart

```diff
@@ -0,0 +1,129 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' as io;
+
+import 'package:args/args.dart';
+import 'package:engine_repo_tools/engine_repo_tools.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as p;
+
+import 'src/header_file.dart';
+
+/// Checks C++ header files for header guards.
+@immutable
+final class HeaderGuardCheck {
+  /// Creates a new header guard checker.
+  const HeaderGuardCheck({
+    required this.source,
+    required this.exclude,
+  });
+
+  /// Parses the command line arguments and creates a new header guard checker.
+  factory HeaderGuardCheck.fromCommandLine(List<String> arguments) {
+    final ArgResults argResults = _parser.parse(arguments);
+    return HeaderGuardCheck(
+      source: Engine.fromSrcPath(argResults['root'] as String),
+      exclude: argResults['exclude'] as List<String>,
+    );
+  }
+
+  /// Engine source root.
+  final Engine source;
+
+  /// Path directories to exclude from the check.
+  final List<String> exclude;
+
+  /// Runs the header guard check.
+  Future<int> run() async {
+    final List<io.File> files = <io.File>[];
+
+  // Recursive search for header files.
+  final io.Directory dir = source.flutterDir;
+  await for (final io.FileSystemEntity entity in dir.list(recursive: true)) {
+    if (entity is io.File && entity.path.endsWith('.h')) {
+      // Check that the file is not excluded.
+      bool excluded = false;
+      for (final String excludePath in exclude) {
+        if (p.isWithin(excludePath, entity.path)) {
+          excluded = true;
+          break;
+        }
+      }
+      if (!excluded) {
+        files.add(entity);
+      }
+    }
+  }
+
+    // Check each file.
+    final List<HeaderFile> badFiles = <HeaderFile>[];
+    for (final io.File file in files) {
```

Comment:
Can we use my package to easily parallelize this? https://pub.dev/packages/pmap

Reply:
Theoretically, though I admit for this example the cost of going async is higher than scanning the text file. I'd be open to it when this takes longer than 1-2s (it currently takes about 1s).